### ISSUE TITLE
Process library test related to sending signals fails on Guix

### ISSUE THREAD

**Issue by Apteryks** [robotframework/robotframework#4292]
Hello!

Running the acceptance test suite (atest) on a Guix System, I have one failure I'm unable to explain:

```
python3 atest/run.py atest/robot/standard_libraries/process         
Python 3.9.9 on Linux
---------------------

Running command:
/gnu/store/6nig0lg4c8qjacwcjpkdxhayqdylk4sq-profile/bin/python3 /tmp/guix-build-python-robotframework-5.0.drv-0/source/src/robot/run.py --doc Robot Framework acceptance tests --metadata interpreter:Python 3.9.9 on Linux --variablefile /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/interpreter.py;/gnu/store/6nig0lg4c8qjacwcjpkdxhayqdylk4sq-profile/bin/python3;Python;3.9.9 --pythonpath /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/resources --outputdir /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/results/Python-3.9.9-Linux --splitlog --console dotted --consolewidth 100 --SuiteStatLevel 3 --exclude require-py3.10 --exclude require-windows atest/robot/standard_libraries/process

Running suite 'Process' with 123 tests.
====================================================================================================
...........................................................................F........................
.......................
----------------------------------------------------------------------------------------------------
FAIL: Process.Sending Signal.By default signal is not sent to process running in shell
Status of 'By default signal is not sent to process running in shell' should have been PASS but it was FAIL.

Error message:
Keyword 'Blastoff Successful' failed after retrying for 2 seconds 200 milliseconds. The last error was: '10
' does not end with 'BLASTOFF'
====================================================================================================
Run suite 'Process' with 123 tests in 58 seconds 132 milliseconds.

FAILED
123 tests, 122 passed, 1 failed

Output:  /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/results/Python-3.9.9-Linux/output.xml
Log:     /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/results/Python-3.9.9-Linux/log.html
Report:  /tmp/guix-build-python-robotframework-5.0.drv-0/source/atest/results/Python-3.9.9-Linux/report.html
```

I've tried to investigate why that is, but don't see anything obvious.  I see that very test is already skipped on MacOSX; why is that?  Does it depend on the shell used for the system?  On Guix System that'd be Bash 5.

Here are the direct dependencies involved, required for the tests mostly:
```
dependencies:
python-docutils@0.17.1
python-invoke@1.6.0
python-jsonschema@3.2.0
python-lxml@4.6.3
python-pygments@2.8.1
python-pyyaml@5.4.1
python-rellu@0.7
python-xmlschema@1.2.5
python@3.9.9
scrot@1.7
tzdata@2021e
xvfb-run@1.20.10-3
```

Any ideas?

**Comment by pekkaklarck** [comment]
I have no idea why this test wouldn't succeed on Guix nor why it was disabled on MacOS. On a brief look at the history the only commit message I found just told the test didn't succeed on OSX. Have you been able to debug this further?

**Comment by Apteryks** [comment]
Hi!  

No, I haven't been able to debug this further.  It fails even outside of the build container, so it's not a problem related to the (lack of) signal handling in the build environment (because there's no PID 1 init there).  With Guix (which can be easily installed atop any GNU/Linux box via [install.sh](https://guix.gnu.org/install.sh)), it should be straightforward to reproduce:

```
$ cp -r $(guix build --source python-robotframework) /tmp/robotframework-5.0 && cd /tmp/robotframework-5.0/
$ chmod -R +rw .
$ guix shell --development python-robotframework strace
[env]$ strace -f -s600 -ostrace.txt ./atest/run.py --test 'Sending Signal.By default signal is not sent to process running in shell' atest/robot/standard_libraries/process/sending_signal.robot
```

The `send_signal_to_process` procedure in Process.py give some clue in its docstring:
```  [...]
        By default sends the signal only to the parent process, not to possible
        child processes started by it. Notice that when `running processes in
        shell`, the shell is the parent process and it depends on the system
        does the shell propagate the signal to the actual started process.
```

**Comment by Apteryks** [comment]
I've spent some time investigating this; and in my case (GNU/Linux with a  kernel Linux 5.16.17-gnu and glibc 2.33 and Python 3.9.9), the Popen with shell=yes ends up doing a fork+exec like this:
```
                    self.pid = _posixsubprocess.fork_exec(
                            args, executable_list,
                            close_fds, tuple(sorted(map(int, fds_to_keep))),
                            cwd, env_list,
                            p2cread, p2cwrite, c2pread, c2pwrite,
                            errread, errwrite,
                            errpipe_read, errpipe_write,
                            restore_signals, start_new_session,
                            gid, gids, uid, umask,
                            preexec_fn)
```
around line 1754 in the `subprocess.py` Python library module. At the debugger, I could see the arguments are populated normally:
```
                    self.pid = _posixsubprocess.fork_exec(
                            args, executable_list,
                            close_fds, tuple(sorted(map(int, fds_to_keep))),
                            cwd, env_list,
                            p2cread, p2cwrite, c2pread, c2pwrite,
                            errread, errwrite,
                            errpipe_read, errpipe_write,
                            restore_signals, start_new_session,
                            gid, gids, uid, umask,
                            preexec_fn)
```
Then after the fork+exec the shell that initiated the command is nowhere to be seen (not in `pstree`, not in `/proc/$PID/cmdline`, so I think it's equivalent to fork+exec, that is, the initial shell process gets replaced by the child it spawns and plays no further role in the execution; indeed the PID returned from Popen is that of the python process spawned, not of the shell that started it.

Shell semantics such as not propagating SIGTERM to child process thus don't hold. I'm not sure why it works on other systems than mine and MacOS.

Anyway, what is this test trying to validate?  That Popen with shell=yes spawns through a shell?  Perhaps the test can simply be removed if we don't know what it is supposed to validate, and why it works sometimes but not always.

**Comment by pekkaklarck** [comment]
It's probably best to run this particular test only on Linux.

**Comment by Apteryks** [comment]
@pekkaklarck hello, and thanks for taking a look; please note that GNU Guix is a GNU/Linux system, so the new condition (system == linux) doesn't help here.  I'd suggest removing the test unless it provides value (so far I don't see any).

**Comment by pekkaklarck** [comment]
Oh, though it would have some other `sys.platform`. My bad, need to think how to handle this. I wouldn't like to remove the test but having `no-quix` tag for excluding it feels overkill.

**Comment by Apteryks** [comment]
> Oh, though it would have some other `sys.platform`. My bad, need to think how to handle this. I wouldn't like to remove the test but having `no-quix` tag for excluding it feels overkill.

There's nothing special about Guix, other than the builds running in a non-networked and minimal chroot, and lacking FHS entries such as `/bin/sh` (which are either patched in or found via other ways such as `PATH`).  It's as plain and well behaved GNU/Linux as you can have :-).  So yeah, a `no-guix` tag doesn't sound right.

I've asked to the very savvy folks in #bash on libera.chat (IRC), and they had to say that Bash will most of the time (not: not always) optimize itself away by exec'ing the command; hence it doesn't stay in the process tree and doesn't handle signals for the process.

In other words, this test erroneously assumes a certain behavior from the shell.  Hence, my recommendation is still to remove it.

**Comment by larryv** [comment]
Hi, I'm one of those #bash miscreants.

When invoked like this...
```sh
sh -c 'python3.9 /tmp/loop.py'
```
...many modern shells (including bash, dash, ksh, yash, and zsh) will `exec` the `python3.9` command as an optimization.

Here is a workaround:
```sh
sh -c 'python3.9 /tmp/loop.py; exit "$?"'
```
This prevents the optimization while preserving `python3.9`'s exit status. It also works with every Bourne-adjacent shell you'll encounter in the wild.

**Comment by pekkaklarck** [comment]
Thanks for the info. My bash knowledge is clearly very limited. :smile: 

The main idea of the test is to validate that by default signals aren't sent to child processes. I would like to keep that but clearly we cannot rely on bash semantics here. I'll check could I create a Python script that starts a child process that I could use instead. If that gets complicated, I'll just remove the test altogether.

**Comment by pekkaklarck** [comment]
And now that I look at the tests I see we already have tests for signals being sent only to the parent process by default. Tests with the shell are basically testing the same exact thing. They can be removed.
